cli/run_util.go

```diff
@@ -42,440 +42,460 @@ import (
 	"github.com/bitrise-io/stepman/toolkits"
 )
 
-func isPRMode(prGlobalFlagPtr *bool, inventoryEnvironments []envmanModels.EnvironmentItemModel) (bool, error) {
-	if prGlobalFlagPtr != nil {
-		return *prGlobalFlagPtr, nil
-	}
+func (r WorkflowRunner) runWorkflow(
+	plan models.WorkflowExecutionPlan,
+	steplibSource string,
+	buildRunResults models.BuildRunResultsModel,
+	environments *[]envmanModels.EnvironmentItemModel, secrets []envmanModels.EnvironmentItemModel,
+	isLastWorkflow bool, tracker analytics.Tracker, buildIDProperties coreanalytics.Properties,
+) models.BuildRunResultsModel {
+	bitrise.PrintRunningWorkflow(plan.WorkflowTitle)
 
-	prIDEnv := os.Getenv(configs.PullRequestIDEnvKey)
-	prModeEnv := os.Getenv(configs.PRModeEnvKey)
+	workflowIDProperties := coreanalytics.Properties{analytics.WorkflowExecutionID: plan.UUID}
+	tracker.SendWorkflowStarted(buildIDProperties.Merge(workflowIDProperties), plan.WorkflowID, plan.WorkflowTitle)
 
-	if prIDEnv != "" || prModeEnv == "true" {
-		return true, nil
+	results := r.activateAndRunSteps(plan, steplibSource, buildRunResults, environments, secrets, isLastWorkflow, tracker, workflowIDProperties)
+
+	tracker.SendWorkflowFinished(workflowIDProperties, results.IsBuildFailed())
+	collectToolVersions(tracker)
+
+	return results
+}
+
+func (r WorkflowRunner) activateAndRunSteps(
+	plan models.WorkflowExecutionPlan,
+	defaultStepLibSource string,
+	buildRunResults models.BuildRunResultsModel,
+	environments *[]envmanModels.EnvironmentItemModel,
+	secrets []envmanModels.EnvironmentItemModel,
+	isLastWorkflow bool,
+	tracker analytics.Tracker,
+	workflowIDProperties coreanalytics.Properties,
+) models.BuildRunResultsModel {
+	log.Debug("[BITRISE_CLI] - Activating and running steps")
+
+	if len(plan.Steps) == 0 {
+		log.Warnf("%s workflow has no steps to run, moving on to the next workflow...", plan.WorkflowTitle)
+		return buildRunResults
 	}
 
-	for _, env := range inventoryEnvironments {
-		key, value, err := env.GetKeyValuePair()
-		if err != nil {
-			return false, err
+	// ------------------------------------------
+	// In function global variables
+
+	// These are global for easy use in local register step run result methods.
+	runResultCollector := newBuildRunResultCollector(tracker)
+
+	// Global variables for synchronising container's lifecycle
+	// and shutting down the last set of containers at the end of the workflow run
+	currentStepGroupID := ""
+	lastStepGroupID := ""
+	defer func() {
```

Comment:
nit: I was looking at the code and noticed this `defer` statement here? With this `defer` it is not easy to see when the `stopContainersForStepGroup` function is called because we only know that it will be called before exiting this function. And as I was looking at the function I noticed that it has a single exit point at the end. 

So my question is that would it make simpler if we kill this defer and move 
```
if lastStepGroupID != "" {
 	r.stopContainersForStepGroup(lastStepGroupID, plan.WorkflowTitle)
 }
```
on line 140 or 144?

Reply:
Updated in https://github.com/bitrise-io/bitrise/pull/969/commits/b4448e2b64684a503aaa35e184c784e2b1b8843d